Switch to use nested tensor by-default in TransformerEncoder

zrphercule opened the pull request:
Summary: Switch to use nested tensor as by default setting in TransformerEncoderLayer.

Test Plan:
CI

Torchtext
buck test mode/opt pytorch/text/test:integration_tests_test_models -- test_xlmr_base_model

Reviewed By: frank-wei

Differential Revision: D36153335

cpuhrsch reviewed (changes requested):
I think we need to be a lot more selective for this optimization, since the padding ratio needs to be relatively high for it to be worth it. Also NestedTensor doesn't autograd for now.

cpuhrsch commented:
cc @jbschlosser @albanD @jisaacso @gchanan

zrphercule2 committed:
Switch to use nested tensor by-default in TransformerEncoder (#77217)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/77217

Switch to use nested tensor as by default setting in TransformerEncoderLayer.

Test Plan:
CI

Torchtext
buck test mode/opt pytorch/text/test:integration_tests_test_models -- test_xlmr_base_model

Reviewed By: frank-wei, cpuhrsch

Differential Revision: D36153335

fbshipit-source-id: 3369e7ff6c193e4367aa90f35f8961315521e0be

albanD commented:
@pytorchbot revert this please as it broke many tests: https://github.com/pytorch/pytorch/runs/6515962633?check_suite_focus=true

albanD commented:
@zrphercule also the PR signal is mostly red. So we should avoid merging PRs with broken CI.
